client/db/src/lib.rs

```diff
@@ -1399,6 +1498,49 @@ impl<Block: BlockT> Backend<Block> {
 
 		Ok(())
 	}
+
+	fn prune_blocks(
+		&self,
+		transaction: &mut Transaction<DbHash>,
+		finalized: NumberFor<Block>,
+	) -> ClientResult<()> {
+		if let KeepBlocks::Some(keep_blocks) = self.keep_blocks {
+			// Always keep the last finalized block
+			let keep = std::cmp::max(keep_blocks, 1);
+			if finalized >= keep.into() {
+				let number = finalized.saturating_sub(keep.into());
+				match read_db(&*self.storage.db, columns::KEY_LOOKUP, columns::BODY, BlockId::<Block>::number(number))? {
+					Some(body) => {
+						utils::remove_db(
+							transaction,
+							&*self.storage.db,
+							columns::KEY_LOOKUP,
+							columns::BODY,
+							BlockId::<Block>::number(number)
+						)?;
+						debug!(target: "db", "Removing block #{}", number);
```

Comment:
Should be moved above the `remove_db` call, because that one could fail and we would not see this log message.